custom_metrics/structured-data.js

```diff
@@ -0,0 +1,269 @@
+//[structured_data]
+// Uncomment the previous line for testing on webpagetest.org
+
+var _logs = [];
+// saves the error details in the results log property.
+// returns the same error object so that it can be also used as the return value for a property.
+function logError(context, messageOrException, exception = null) {
+  let error = { type: "error", context: context };
+
+  try {
+    if (typeof messageOrException === "string") {
+      error.message = messageOrException;
+    } else if (messageOrException instanceof Object) {
+      error.exception = messageOrException;
+
+      if (messageOrException.message) {
+        error.message = messageOrException.message;
+      } else {
+        error.message = JSON.stringify(
+          messageOrException,
+          Object.getOwnPropertyNames(messageOrException)
+        );
+      }
+    } else {
+      error.message = JSON.stringify(messageOrException);
+    }
+
+    if (exception) {
+      error.exception = exception;
+
+      if (exception.message) {
+        error.message += ": " + exception.message;
+      }
+    }
+  } catch (e) {
+    error.message = "logError failed";
+    error.exception = e;
+  }
+
+  _logs.push(error);
+
+  return error;
+}
+
+var _custom_metrics = null;
+
+try {
+  // whole process is placed in a try/catch so we can log uncaught errors
+
+  // this provides access to a lot of WebPageTest data, including the raw html, headers and other requests involved
+  var _wptBodies = [];
+  try {
+    _wptBodies = $WPT_BODIES;
+  } catch (e) {
+    logError("wptBodies", "Data returned was not valid", e);
+  }
+
+  // string of the raw html
+  let _rawHtml = null;
+  function getRawHtml() {
+    if (!_rawHtml && _wptBodies.length > 0) {
+      _rawHtml = _wptBodies[0].response_body;
+    }
+    return _rawHtml;
+  }
+
+  // an html document of the raw html
+  let _rawHtmlDocument = null;
+  function getRawHtmlDocument() {
+    if (!_rawHtmlDocument) {
+      let html = getRawHtml();
+
+      if (html) {
+        _rawHtmlDocument =
+          document.implementation.createHTMLDocument("New Document");
+
+        _rawHtmlDocument.documentElement.innerHTML = html;
+      }
+    }
+    return _rawHtmlDocument;
+  }
+
+  _custom_metrics = {
+    structured_data: (() => {
+      try {
+        function dumpTagAttributes(tags) {
+          return tags.map((tag) => {
+            const json_tag = {};
+            ["rel", "href", "name", "property", "content"].map((attribute) => {
```

Comment:
Nit: not actually mapping the results
```suggestion
            ["rel", "href", "name", "property", "content"].forEach((attribute) => {
```